node/util_test.go

```diff
@@ -0,0 +1,153 @@
+package node
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/config"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// NewNetwork creates a network of fully connected nodes.
+func NewNetwork(conf config.Config, l log.Log, size int) ([]*TestApp, func() error, error) {
+	// We need to set this global state
+	types.SetLayersPerEpoch(conf.LayersPerEpoch)
+	types.SetNetworkHRP(conf.NetworkHRP) // set to generate coinbase
+
+	// To save an epoch of startup time, we bootstrap (meaning we manually set
+	// it) the beacon for epoch 2 so that in epoch 3 hare can start.
+	bootstrapEpoch := (types.GetEffectiveGenesis() + 1).GetEpoch()
+	bootstrapBeacon := types.Beacon{}
+	genesis := conf.Genesis.GenesisID()
+	copy(bootstrapBeacon[:], genesis[:])
+
+	ctx, cancel := context.WithCancel(context.Background())
+	g := errgroup.Group{}
+	var apps []*TestApp
+	var datadirs []string
+	cleanup := func() error {
+		cancel()
+		// Wait for nodes to shutdown
+		g.Wait()
+		// Clean their datadirs
+		for _, d := range datadirs {
+			err := os.RemoveAll(d)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// We encode and decode the config in order to deep copy it.
+	var buf bytes.Buffer
+	err := gob.NewEncoder(&buf).Encode(conf)
+	if err != nil {
+		cleanup()
+		return nil, nil, err
+	}
+	marshaled := buf.Bytes()
+
+	for i := 0; i < size; i++ {
+		var c config.Config
+		err := gob.NewDecoder(bytes.NewBuffer(marshaled)).Decode(&c)
```

Comment:
```suggestion
		c := conf
```

Encoding / decoding is not necessary here. All values that are changed between the copies are value types and the service doesn't alter the config that is passed into it.

Reply:
I'm happy to change this, it will remove the need to modify PostProviderID, however it seems like it will be hard to enforce going forward since there doesn't seem to be any written standard around how to handle config. Let me know your thoughts?

Reply:
So given the problem exposing the internals of PostProviderID, I've decided to remove the deep copy.